src/nanoemoji/bitmap_tables.py

```diff
@@ -0,0 +1,269 @@
+            ),
+            line_height=line_height,
+            line_ascent=int(line_ascent),
+        )
+
+        print("line_height", metrics.line_height)
+        print("line_ascent", metrics.line_ascent)
+        print("metrics.x_offset", metrics.x_offset)
+        print("metrics.y_offset", metrics.y_offset)
+
+        return metrics
+
+
+# https://github.com/googlefonts/noto-emoji/blob/9a5261d871451f9b5183c93483cbd68ed916b1e9/third_party/color_emoji/emoji_builder.py#L53
+def _ppem(config: FontConfig, advance: int) -> int:
+    return int(config.bitmap_resolution * config.upem / advance)
+
+
+def _advance(ttfont: ttLib.TTFont, color_glyphs: Sequence[ColorGlyph]) -> int:
+    # let's go ahead and fail miserably if advances are not all the same
+    # proportional bitmaps can wait for a second pass :)
+    advances = {
+        ttfont["hmtx"].metrics[ttfont.getGlyphName(c.glyph_id)][0] for c in color_glyphs
+    }
+    assert len(advances) == 1, "Proportional bitmaps not supported yet"
+    return next(iter(advances))
+
+
+def _image_bytes(color_glyph: ColorGlyph) -> bytes:
+    with open(color_glyph.filename, "rb") as f:
+        return f.read()
+
+
+def _cbdt_record_size(image_format: int, image_data: bytes) -> int:
+    assert image_format == _CBDT_SMALL_METRIC_PNGS, "Unrecognized format"
+    return _CBDT_SMALL_METRIC_PNG_HEADER_SIZE + len(image_data)
+
+
+def _cbdt_bitmapdata_offsets(
+    image_format: int, color_glyphs: Sequence[ColorGlyph]
+) -> List[Tuple[int, int]]:
+    # TODO is this right? - feels dumb. But ... compile crashes if locations are unknown.
+    offsets = []
+    offset = _CBDT_HEADER_SIZE
+    for color_glyph in color_glyphs:
+        offsets.append(offset)
+        offset += _cbdt_record_size(image_format, _image_bytes(color_glyph))
+    offsets.append(offset)  # capture end of stream
+    return list(zip(offsets, offsets[1:]))
+
+
+def _cbdt_bitmap_data(
+    config: FontConfig, metrics: BitmapMetrics, image_data: bytes
+) -> CbdtBitmapFormat17:
+
+    bitmap_data = CbdtBitmapFormat17(b"", None)
+    bitmap_data.metrics = SmallGlyphMetrics()
+    bitmap_data.metrics.height = config.bitmap_resolution
+    bitmap_data.metrics.width = config.bitmap_resolution
+    # center within advance
+    bitmap_data.metrics.BearingX = metrics.x_offset
+    bitmap_data.metrics.BearingY = metrics.y_offset
+    bitmap_data.metrics.Advance = config.bitmap_advance
+    bitmap_data.imageData = image_data
+    return bitmap_data
+
+
+def make_sbix_table(
+    config: FontConfig,
+    ttfont: ttLib.TTFont,
+    color_glyphs: Sequence[ColorGlyph],
+):
+
+    sbix = ttLib.newTable("sbix")
+    ttfont[sbix.tableTag] = sbix
+
+    sbix.ppem = _ppem(config, _advance(ttfont, color_glyphs))
+    # TODO what goes here
+    sbix.ppi = 96
+
+    strike = SbixStrike()
+    strike.ppem = sbix.ppem + 8  # TODO magic #
+    strike.resolution = config.bitmap_resolution
```

Comment:
Apple Color Emoji has 72 for this field, we should do the same.

With the patch below, I can finally see the correct size/placement in Safari Mac for the sbix font:

```diff
diff --git a/src/nanoemoji/bitmap_tables.py b/src/nanoemoji/bitmap_tables.py
index 61b8353..52d9e00 100644
--- a/src/nanoemoji/bitmap_tables.py
+++ b/src/nanoemoji/bitmap_tables.py
@@ -85,7 +85,7 @@ class BitmapMetrics(NamedTuple):
 
 # https://github.com/googlefonts/noto-emoji/blob/9a5261d871451f9b5183c93483cbd68ed916b1e9/third_party/color_emoji/emoji_builder.py#L53
 def _ppem(config: FontConfig, advance: int) -> int:
-    return int(config.bitmap_resolution * config.upem / advance)
+    return int(config.bitmap_advance * config.upem / advance)
 
 
 def _advance(ttfont: ttLib.TTFont, color_glyphs: Sequence[ColorGlyph]) -> int:
@@ -146,13 +146,12 @@ def make_sbix_table(
     sbix = ttLib.newTable("sbix")
     ttfont[sbix.tableTag] = sbix
 
-    sbix.ppem = _ppem(config, _advance(ttfont, color_glyphs))
+    ppem = _ppem(config, _advance(ttfont, color_glyphs))
     # TODO what goes here
-    sbix.ppi = 96
 
     strike = SbixStrike()
-    strike.ppem = sbix.ppem + 8  # TODO magic #
-    strike.resolution = config.bitmap_resolution
+    strike.ppem = ppem
+    strike.resolution = 72  # pixels per inch
     sbix.strikes[strike.ppem] = strike
 
     metrics = BitmapMetrics.create(config, strike.ppem)
@@ -167,7 +166,7 @@ def make_sbix_table(
             glyphName=glyph_name,
             imageData=image_data,
             originOffsetX=metrics.x_offset,
-            originOffsetY=metrics.line_height - metrics.y_offset,
+            originOffsetY=metrics.line_ascent - metrics.line_height,
         )
         strike.glyphs[glyph_name] = glyph
 
@@ -189,7 +188,7 @@ def make_cbdt_table(
     ), "Below assumes color gyphs gids are consecutive"
 
     advance = _advance(ttfont, color_glyphs)
-    ppem = _ppem(config, advance) + 8
+    ppem = _ppem(config, advance)
 
     cbdt = ttLib.newTable("CBDT")
     ttfont[cbdt.tableTag] = cbdt
```

![image](https://user-images.githubusercontent.com/6939968/150563540-0d376c65-152e-4ca1-a9f5-ac73b4ef4785.png)

Reply:
Updated so but ofc ptal and confirm it looks sane

Reply:
Adopted the final change (originOffsetY) based on discussions such as https://github.com/harfbuzz/harfbuzz/issues/2679#issuecomment-1021419864. I suppose I shall have to abandon the notion that sbix is the bitmap format that works in the most places.